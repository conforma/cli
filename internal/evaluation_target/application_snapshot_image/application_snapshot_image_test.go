// Copyright 2022 Red Hat, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

//go:build unit

package application_snapshot_image

import (
	"context"
	_ "embed"
	"encoding/base64"
	"encoding/json"
	"errors"
	"regexp"
	"strings"
	"testing"

	"github.com/google/go-containerregistry/pkg/name"
	v1 "github.com/google/go-containerregistry/pkg/v1"
	"github.com/google/go-containerregistry/pkg/v1/remote"
	"github.com/in-toto/in-toto-golang/in_toto"
	"github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/common"
	v02 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
	"github.com/secure-systems-lab/go-securesystemslib/dsse"
	"github.com/sigstore/cosign/v2/pkg/cosign"
	"github.com/sigstore/cosign/v2/pkg/oci"
	"github.com/sigstore/cosign/v2/pkg/oci/static"
	"github.com/spf13/afero"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"

	"github.com/enterprise-contract/ec-cli/internal/attestation"
	"github.com/enterprise-contract/ec-cli/internal/evaluator"
	"github.com/enterprise-contract/ec-cli/internal/mocks"
	"github.com/enterprise-contract/ec-cli/internal/output"
	"github.com/enterprise-contract/ec-cli/internal/policy"
	"github.com/enterprise-contract/ec-cli/internal/sigstore"
	"github.com/enterprise-contract/ec-cli/internal/utils"
)

// pipelineRunBuildType is the type of attestation we're interested in evaluating
const pipelineRunBuildType = "https://tekton.dev/attestations/chains/pipelinerun@v2"

func TestNewApplicationSnapshotImage(t *testing.T) {
	ctx := context.Background()
	p, err := policy.NewOfflinePolicy(ctx, policy.Now)
	require.NoError(t, err)

	asi, err := NewApplicationSnapshotImage(ctx, "example.com/test:latest", p)
	require.NoError(t, err)

	assert.Len(t, asi.Validators, 1)
}

func TestApplicationSnapshotImage_ValidateImageAccess(t *testing.T) {
	type fields struct {
		reference    name.Reference
		checkOpts    cosign.CheckOpts
		attestations []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]
		Evaluator    evaluator.Evaluator
	}
	type args struct {
		ctx context.Context
	}
	ref, _ := name.ParseReference("registry/image:tag")
	tests := []struct {
		name    string
		fields  fields
		args    args
		wantErr bool
	}{
		{
			name: "Returns no error when able to access image ref",
			fields: fields{
				reference:    ref,
				checkOpts:    cosign.CheckOpts{},
				attestations: nil,
				Evaluator:    nil,
			},
			args:    args{ctx: context.Background()},
			wantErr: false,
		},
		{
			name: "Returns error when unable to access image ref",
			fields: fields{
				reference:    ref,
				checkOpts:    cosign.CheckOpts{},
				attestations: nil,
				Evaluator:    nil,
			},
			args:    args{ctx: context.Background()},
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.wantErr {
				imageRefTransport = remote.WithTransport(&mocks.HttpTransportMockFailure{})
			} else {
				imageRefTransport = remote.WithTransport(&mocks.HttpTransportMockSuccess{})
			}
			a := &ApplicationSnapshotImage{
				reference:    tt.fields.reference,
				checkOpts:    tt.fields.checkOpts,
				attestations: tt.fields.attestations,
				Evaluators:   []evaluator.Evaluator{tt.fields.Evaluator},
			}
			if err := a.ValidateImageAccess(tt.args.ctx); (err != nil) != tt.wantErr {
				t.Errorf("ValidateImageAccess() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

type fakeAtt struct {
	statement in_toto.ProvenanceStatementSLSA02
}

func (f fakeAtt) Data() []byte {
	bytes, err := json.Marshal(f.statement)
	if err != nil {
		panic(err)
	}
	return bytes
}

func (f fakeAtt) Statement() in_toto.ProvenanceStatementSLSA02 {
	return f.statement
}

func (f fakeAtt) Signatures() []output.EntitySignature {
	return nil
}

func createSimpleAttestation(statement *in_toto.ProvenanceStatementSLSA02) attestation.Attestation[in_toto.ProvenanceStatementSLSA02] {
	if statement == nil {
		statement = &in_toto.ProvenanceStatementSLSA02{
			StatementHeader: in_toto.StatementHeader{
				Type:          in_toto.StatementInTotoV01,
				PredicateType: v02.PredicateSLSAProvenance,
			},
			Predicate: v02.ProvenancePredicate{
				BuildType: pipelineRunBuildType,
			},
		}
	}

	return fakeAtt{statement: *statement}
}

const simpleAttestationJSONText = `{
	"_type": "https://in-toto.io/Statement/v0.1",
	"predicateType": "https://slsa.dev/provenance/v0.2",
	"subject": null,
	"predicate": {
		"buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
		"builder": {
			"id": ""
		},
		"invocation": {
			"configSource": {}
		}
	}
}`

func TestWriteInputFile(t *testing.T) {
	cases := []struct {
		name     string
		snapshot ApplicationSnapshotImage
		want     string
	}{
		{
			name: "single attestations",
			snapshot: ApplicationSnapshotImage{
				reference:    name.MustParseReference("registry.io/repository/image:tag"),
				attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{createSimpleAttestation(nil)},
			},
			want: `{"attestations": [` + simpleAttestationJSONText + `], "image": {"ref": "registry.io/repository/image:tag"}}`,
		},
		{
			name: "multiple attestations",
			snapshot: ApplicationSnapshotImage{
				reference: name.MustParseReference("registry.io/repository/image:tag"),
				attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
					createSimpleAttestation(nil),
					createSimpleAttestation(nil),
				},
			},
			want: `{"attestations": [` + simpleAttestationJSONText + "," + simpleAttestationJSONText + `], "image": {"ref": "registry.io/repository/image:tag"}}`,
		},
		{
			name: "image signatures",
			snapshot: ApplicationSnapshotImage{
				reference: name.MustParseReference("registry.io/repository/image:tag"),
				signatures: []output.EntitySignature{
					{
						KeyID:     "keyId1",
						Signature: "signature1",
						Chain:     []string{"certificate1", "certificate2"},
					},
					{
						KeyID:     "keyId2",
						Signature: "signature2",
					},
				},
				attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
					createSimpleAttestation(nil),
				},
			},
			want: `{"attestations": [` + simpleAttestationJSONText + `], "image": {"ref": "registry.io/repository/image:tag", "signatures": [{"keyid": "keyId1", "sig": "signature1", "chain": ["certificate1", "certificate2"]}, {"keyid": "keyId2", "sig": "signature2"}]}}`,
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			fs := afero.NewMemMapFs()
			ctx := utils.WithFS(context.Background(), fs)
			input, err := tt.snapshot.WriteInputFile(ctx)

			assert.NoError(t, err)
			assert.NotEmpty(t, input)
			assert.Regexp(t, `/ecp_input.\d+/input.json`, input)
			fileExists, err := afero.Exists(fs, input)
			assert.NoError(t, err)
			assert.True(t, fileExists)

			bytes, err := afero.ReadFile(fs, input)
			assert.NoError(t, err)
			assert.JSONEq(t, tt.want, string(bytes))
		})
	}
}

func TestWriteInputFileMultipleAttestations(t *testing.T) {
	att := createSimpleAttestation(nil)
	a := ApplicationSnapshotImage{
		reference:    name.MustParseReference("registry.io/repository/image:tag"),
		attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{att},
	}

	fs := afero.NewMemMapFs()
	ctx := utils.WithFS(context.Background(), fs)
	input, err := a.WriteInputFile(ctx)

	assert.NoError(t, err)
	assert.NotEmpty(t, input)
	assert.Regexp(t, `/ecp_input.\d+/input.json`, input)
	fileExists, err := afero.Exists(fs, input)
	assert.NoError(t, err)
	assert.True(t, fileExists)

	bytes, err := afero.ReadFile(fs, input)
	assert.NoError(t, err)
	assert.JSONEq(t, `{"attestations": [`+simpleAttestationJSONText+`], "image": {"ref": "registry.io/repository/image:tag"}}`, string(bytes))
}

func TestSyntaxValidationWithoutAttestations(t *testing.T) {
	noAttestations := ApplicationSnapshotImage{}

	err := noAttestations.ValidateAttestationSyntax(context.TODO())
	assert.Error(t, err, "Expected error in validation")

	assert.True(t, strings.HasPrefix(err.Error(), "EV001: No attestation data"))
}

func TestSyntaxValidation(t *testing.T) {
	valid := createSimpleAttestation(&in_toto.ProvenanceStatementSLSA02{
		StatementHeader: in_toto.StatementHeader{
			Type:          in_toto.StatementInTotoV01,
			PredicateType: v02.PredicateSLSAProvenance,
			Subject: []in_toto.Subject{
				{
					Name: "hello",
					Digest: common.DigestSet{
						"sha1": "abcdef0123456789",
					},
				},
			},
		},
		Predicate: v02.ProvenancePredicate{
			BuildType: pipelineRunBuildType,
			Builder: common.ProvenanceBuilder{
				ID: "scheme:uri",
			},
		},
	})

	invalid := createSimpleAttestation(&in_toto.ProvenanceStatementSLSA02{
		StatementHeader: in_toto.StatementHeader{
			Type:          in_toto.StatementInTotoV01,
			PredicateType: v02.PredicateSLSAProvenance,
			Subject: []in_toto.Subject{
				{
					Name: "hello",
					Digest: common.DigestSet{
						"sha1": "abcdef0123456789",
					},
				},
			},
		},
		Predicate: v02.ProvenancePredicate{
			BuildType: pipelineRunBuildType,
			Builder: common.ProvenanceBuilder{
				ID: "invalid", // must be in URI syntax
			},
		},
	})

	cases := []struct {
		name         string
		attestations []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]
		err          *regexp.Regexp
	}{
		{
			name: "invalid",
			attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
				invalid,
			},
			err: regexp.MustCompile(`EV003: Attestation syntax validation failed, .*, caused by:\nSchema ID: https://slsa.dev/provenance/v0.2\n - /predicate/builder/id: "invalid" invalid uri: uri missing scheme prefix`),
		},
		{
			name: "valid",
			attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
				valid,
			},
		},
		{
			name: "empty",
			attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
				createSimpleAttestation(&in_toto.ProvenanceStatementSLSA02{}),
			},
			err: regexp.MustCompile(`EV002: Unable to decode attestation data from attestation image, .*, caused by: unexpected end of JSON input`),
		},
		{
			name: "valid and invalid",
			attestations: []attestation.Attestation[in_toto.ProvenanceStatementSLSA02]{
				valid,
				invalid,
			},
			err: regexp.MustCompile(`EV003`),
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			a := ApplicationSnapshotImage{
				attestations: c.attestations,
			}

			err := a.ValidateAttestationSyntax(context.TODO())
			if c.err == nil {
				assert.NoError(t, err)
			} else {
				assert.Error(t, err)
				assert.Regexp(t, err, err.Error())
			}
		})
	}
}

type MockClient struct {
	mock.Mock
}

func (c *MockClient) VerifyImageSignatures(ctx context.Context, name name.Reference, opts *cosign.CheckOpts) ([]oci.Signature, bool, error) {
	args := c.Called(ctx, name, opts)

	return args.Get(0).([]oci.Signature), args.Get(1).(bool), args.Error(2)
}

func (c *MockClient) VerifyImageAttestations(ctx context.Context, name name.Reference, opts *cosign.CheckOpts) ([]oci.Signature, bool, error) {
	args := c.Called(ctx, name, opts)

	return args.Get(0).([]oci.Signature), args.Get(1).(bool), args.Error(2)
}

func (c *MockClient) Head(name name.Reference, options ...remote.Option) (*v1.Descriptor, error) {
	args := c.Called(name, options)

	return args.Get(0).(*v1.Descriptor), args.Error(1)
}

func TestValidateAttestationSignatureClaims(t *testing.T) {
	ref := name.MustParseReference("registry.io/repository/image:tag")
	a := ApplicationSnapshotImage{
		reference: ref,
	}

	c := MockClient{}

	ctx := WithClient(context.Background(), &c)
	ctx = sigstore.WithClient(ctx, &c)

	c.On("VerifyImageAttestations", ctx, ref, mock.Anything).Return([]oci.Signature{}, false, nil)

	err := a.ValidateAttestationSignature(ctx)
	require.NoError(t, err)

	call := c.Calls[0]

	checkOpts := call.Arguments.Get(2).(*cosign.CheckOpts)
	assert.NotNil(t, checkOpts)

	claimVerifier := checkOpts.ClaimVerifier
	assert.NotNil(t, claimVerifier)

	cases := []struct {
		name      string
		statement in_toto.Statement
		digest    v1.Hash
		err       error
	}{
		{
			name: "happy day",
			statement: in_toto.Statement{
				StatementHeader: in_toto.StatementHeader{
					Subject: []in_toto.Subject{
						{
							Digest: map[string]string{
								"sha256": "dabbad00",
							},
						},
					},
				},
			},
			digest: v1.Hash{Algorithm: "sha256", Hex: "dabbad00"},
		},
		{
			name: "happy day - multiple digests",
			statement: in_toto.Statement{
				StatementHeader: in_toto.StatementHeader{
					Subject: []in_toto.Subject{
						{
							Digest: map[string]string{
								"sha512": "dead10cc",
								"sha256": "dabbad00",
							},
						},
					},
				},
			},
			digest: v1.Hash{Algorithm: "sha256", Hex: "dabbad00"},
		},
		{
			name: "no digests",
			statement: in_toto.Statement{
				StatementHeader: in_toto.StatementHeader{
					Subject: []in_toto.Subject{},
				},
			},
			digest: v1.Hash{Algorithm: "sha256", Hex: "dabbad00"},
			err:    errors.New("no matching subject digest found"),
		},
		{
			name: "mismatched digests",
			statement: in_toto.Statement{
				StatementHeader: in_toto.StatementHeader{
					Subject: []in_toto.Subject{
						{
							Digest: map[string]string{
								"sha256": "dead10cc",
							},
						},
					},
				},
			},
			digest: v1.Hash{Algorithm: "sha256", Hex: "dabbad00"},
			err:    errors.New("no matching subject digest found"),
		},
		{
			name:      "empty statement",
			statement: in_toto.Statement{},
			digest:    v1.Hash{Algorithm: "sha256", Hex: "dabbad00"},
			err:       errors.New("no matching subject digest found"),
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			statementJSON, err := json.Marshal(c.statement)
			require.NoError(t, err)

			statement := base64.StdEncoding.EncodeToString(statementJSON)

			dsse := dsse.Envelope{
				Payload: statement,
			}

			payload, err := json.Marshal(dsse)
			require.NoError(t, err)

			signature, err := static.NewSignature(payload, "signature")
			require.NoError(t, err)

			err = claimVerifier(signature, c.digest, nil)
			assert.Equal(t, c.err, err)
		})
	}
}

func TestAddSignatures(t *testing.T) {
	a := ApplicationSnapshotImage{}
	assert.Empty(t, a.Signatures())
	a.AddSignatures(output.EntitySignature{})
	assert.Len(t, a.Signatures(), 1)
	a.AddSignatures(output.EntitySignature{})
	assert.Len(t, a.Signatures(), 2)
}
