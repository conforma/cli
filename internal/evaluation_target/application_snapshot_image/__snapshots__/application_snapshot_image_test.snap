
[TestValidateImageSignatureWithCertificates - 1]
[]signature.EntitySignature{
    {
        KeyID:       "6add046e38418d021a562c6a8633d5eca7379595",
        Signature:   "signature",
        Certificate: "-----BEGIN CERTIFICATE-----\nMIIG2TCCBl+gAwIBAgIUdtQgx3Mj6A3T0X7Oh8bS1nNABTEwCgYIKoZIzj0EAwMw\nNzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl\ncm1lZGlhdGUwHhcNMjMwNjA3MDMxNDEyWhcNMjMwNjA3MDMyNDEyWjAAMFkwEwYH\nKoZIzj0CAQYIKoZIzj0DAQcDQgAEz6tsPZHx7njElmbGbMYxKiYneuofINbOE8Tg\n1gkyQcckWyu1xA/Fs0O1SpPkn/KJYLJ3J5ziqgd1EguuCqK3Z6OCBX4wggV6MA4G\nA1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQUat0E\nbjhBjQIaVixqhjPV7Kc3lZUwHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y\nZD8waAYDVR0RAQH/BF4wXIZaaHR0cHM6Ly9naXRodWIuY29tL2NoYWluZ3VhcmQt\naW1hZ2VzL2ltYWdlcy8uZ2l0aHViL3dvcmtmbG93cy9yZWxlYXNlLnlhbWxAcmVm\ncy9oZWFkcy9tYWluMDkGCisGAQQBg78wAQEEK2h0dHBzOi8vdG9rZW4uYWN0aW9u\ncy5naXRodWJ1c2VyY29udGVudC5jb20wEgYKKwYBBAGDvzABAgQEcHVzaDA2Bgor\nBgEEAYO/MAEDBChlMWRjZGY3MGJlMzI2YTQ5NDI5NTc1NDYyMmZlMzQ2MzE2MDA1\nMzFhMCwGCisGAQQBg78wAQQEHi5naXRodWIvd29ya2Zsb3dzL3JlbGVhc2UueWFt\nbDAmBgorBgEEAYO/MAEFBBhjaGFpbmd1YXJkLWltYWdlcy9pbWFnZXMwHQYKKwYB\nBAGDvzABBgQPcmVmcy9oZWFkcy9tYWluMDsGCisGAQQBg78wAQgELQwraHR0cHM6\nLy90b2tlbi5hY3Rpb25zLmdpdGh1YnVzZXJjb250ZW50LmNvbTBqBgorBgEEAYO/\nMAEJBFwMWmh0dHBzOi8vZ2l0aHViLmNvbS9jaGFpbmd1YXJkLWltYWdlcy9pbWFn\nZXMvLmdpdGh1Yi93b3JrZmxvd3MvcmVsZWFzZS55YW1sQHJlZnMvaGVhZHMvbWFp\nbjA4BgorBgEEAYO/MAEKBCoMKGUxZGNkZjcwYmUzMjZhNDk0Mjk1NzU0NjIyZmUz\nNDYzMTYwMDUzMWEwHQYKKwYBBAGDvzABCwQPDA1naXRodWItaG9zdGVkMDsGCisG\nAQQBg78wAQwELQwraHR0cHM6Ly9naXRodWIuY29tL2NoYWluZ3VhcmQtaW1hZ2Vz\nL2ltYWdlczA4BgorBgEEAYO/MAENBCoMKGUxZGNkZjcwYmUzMjZhNDk0Mjk1NzU0\nNjIyZmUzNDYzMTYwMDUzMWEwHwYKKwYBBAGDvzABDgQRDA9yZWZzL2hlYWRzL21h\naW4wGQYKKwYBBAGDvzABDwQLDAk1NjM1MTA5NTIwNAYKKwYBBAGDvzABEAQmDCRo\ndHRwczovL2dpdGh1Yi5jb20vY2hhaW5ndWFyZC1pbWFnZXMwGQYKKwYBBAGDvzAB\nEQQLDAkxMTMxOTg1NDUwagYKKwYBBAGDvzABEgRcDFpodHRwczovL2dpdGh1Yi5j\nb20vY2hhaW5ndWFyZC1pbWFnZXMvaW1hZ2VzLy5naXRodWIvd29ya2Zsb3dzL3Jl\nbGVhc2UueWFtbEByZWZzL2hlYWRzL21haW4wOAYKKwYBBAGDvzABEwQqDChlMWRj\nZGY3MGJlMzI2YTQ5NDI5NTc1NDYyMmZlMzQ2MzE2MDA1MzFhMBQGCisGAQQBg78w\nARQEBgwEcHVzaDBeBgorBgEEAYO/MAEVBFAMTmh0dHBzOi8vZ2l0aHViLmNvbS9j\naGFpbmd1YXJkLWltYWdlcy9pbWFnZXMvYWN0aW9ucy9ydW5zLzUxOTU1MDc2MzYv\nYXR0ZW1wdHMvMTCBigYKKwYBBAHWeQIEAgR8BHoAeAB2AN09MGrGxxEyYxkeHJln\nNwKiSl643jyt/4eKcoAvKe6OAAABiJPZADAAAAQDAEcwRQIgdHXB0QGS/GWkBnY1\nAZXSwb6/tbnnaVeWzde3t0fkkRMCIQC0bwdhWep548Cp4LzBPgGD0eioadqQdJHe\nXtVXBkD1dDAKBggqhkjOPQQDAwNoADBlAjBPpXDUSaAk5D6T1Eaqh+TRSQXr6rqV\nYxAJb/NgDbq8tTVLKustJDu2V9TQcpSzuKICMQDt0EAHmTISmKC8H3dciTrySh2l\nuS2rfl+L2AFS6DxAmVTBR3dlbrxQsUxshBWyH5s=\n-----END CERTIFICATE-----\n",
        Chain:       {"-----BEGIN CERTIFICATE-----\nMIICGjCCAaGgAwIBAgIUALnViVfnU0brJasmRkHrn/UnfaQwCgYIKoZIzj0EAwMw\nKjEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MREwDwYDVQQDEwhzaWdzdG9yZTAeFw0y\nMjA0MTMyMDA2MTVaFw0zMTEwMDUxMzU2NThaMDcxFTATBgNVBAoTDHNpZ3N0b3Jl\nLmRldjEeMBwGA1UEAxMVc2lnc3RvcmUtaW50ZXJtZWRpYXRlMHYwEAYHKoZIzj0C\nAQYFK4EEACIDYgAE8RVS/ysH+NOvuDZyPIZtilgUF9NlarYpAd9HP1vBBH1U5CV7\n7LSS7s0ZiH4nE7Hv7ptS6LvvR/STk798LVgMzLlJ4HeIfF3tHSaexLcYpSASr1kS\n0N/RgBJz/9jWCiXno3sweTAOBgNVHQ8BAf8EBAMCAQYwEwYDVR0lBAwwCgYIKwYB\nBQUHAwMwEgYDVR0TAQH/BAgwBgEB/wIBADAdBgNVHQ4EFgQU39Ppz1YkEZb5qNjp\nKFWixi4YZD8wHwYDVR0jBBgwFoAUWMAeX5FFpWapesyQoZMi0CrFxfowCgYIKoZI\nzj0EAwMDZwAwZAIwPCsQK4DYiZYDPIaDi5HFKnfxXx6ASSVmERfsynYBiX2X6SJR\nnZU84/9DZdnFvvxmAjBOt6QpBlc4J/0DxvkTCqpclvziL6BCCPnjdlIB3Pu3BxsP\nmygUY7Ii2zbdCdliiow=\n-----END CERTIFICATE-----\n", "-----BEGIN CERTIFICATE-----\nMIIB9zCCAXygAwIBAgIUALZNAPFdxHPwjeDloDwyYChAO/4wCgYIKoZIzj0EAwMw\nKjEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MREwDwYDVQQDEwhzaWdzdG9yZTAeFw0y\nMTEwMDcxMzU2NTlaFw0zMTEwMDUxMzU2NThaMCoxFTATBgNVBAoTDHNpZ3N0b3Jl\nLmRldjERMA8GA1UEAxMIc2lnc3RvcmUwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAAT7\nXeFT4rb3PQGwS4IajtLk3/OlnpgangaBclYpsYBr5i+4ynB07ceb3LP0OIOZdxex\nX69c5iVuyJRQ+Hz05yi+UF3uBWAlHpiS5sh0+H2GHE7SXrk1EC5m1Tr19L9gg92j\nYzBhMA4GA1UdDwEB/wQEAwIBBjAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBRY\nwB5fkUWlZql6zJChkyLQKsXF+jAfBgNVHSMEGDAWgBRYwB5fkUWlZql6zJChkyLQ\nKsXF+jAKBggqhkjOPQQDAwNpADBmAjEAj1nHeXZp+13NWBNa+EDsDP8G1WWg1tCM\nWP/WHPqpaVo0jhsweNFZgSs0eE7wYI4qAjEA2WB9ot98sIkoF3vZYdd3/VtWB5b9\nTNMea7Ix/stJ5TfcLLeABLE4BNJOsQ4vnBHJ\n-----END CERTIFICATE-----\n"},
        Metadata:    {"Fulcio Build Config Digest":"e1dcdf70be326a494295754622fe34631600531a", "Fulcio Build Config URI":"https://github.com/chainguard-images/images/.github/workflows/release.yaml@refs/heads/main", "Fulcio Build Signer Digest":"e1dcdf70be326a494295754622fe34631600531a", "Fulcio Build Signer URI":"https://github.com/chainguard-images/images/.github/workflows/release.yaml@refs/heads/main", "Fulcio Build Trigger":"push", "Fulcio GitHub Workflow Name":".github/workflows/release.yaml", "Fulcio GitHub Workflow Ref":"refs/heads/main", "Fulcio GitHub Workflow Repository":"chainguard-images/images", "Fulcio GitHub Workflow SHA":"e1dcdf70be326a494295754622fe34631600531a", "Fulcio GitHub Workflow Trigger":"push", "Fulcio Issuer":"https://token.actions.githubusercontent.com", "Fulcio Issuer (V2)":"https://token.actions.githubusercontent.com", "Fulcio Run Invocation URI":"https://github.com/chainguard-images/images/actions/runs/5195507636/attempts/1", "Fulcio Runner Environment":"github-hosted", "Fulcio Source Repository Digest":"e1dcdf70be326a494295754622fe34631600531a", "Fulcio Source Repository Identifier":"563510952", "Fulcio Source Repository Owner Identifier":"113198545", "Fulcio Source Repository Owner URI":"https://github.com/chainguard-images", "Fulcio Source Repository Ref":"refs/heads/main", "Fulcio Source Repository URI":"https://github.com/chainguard-images/images", "Issuer":"CN=sigstore-intermediate,O=sigstore.dev", "Not After":"2023-06-07T03:24:12Z", "Not Before":"2023-06-07T03:14:12Z", "Serial Number":"76d420c77323e80dd3d17ece87c6d2d673400531", "Subject Alternative Name":"URIs:https://github.com/chainguard-images/images/.github/workflows/release.yaml@refs/heads/main"},
    },
}
---

[TestWriteInputFile/single_attestations - 1]
{
 "attestations": [
  {
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  }
 ],
 "image": {
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFile/multiple_attestations - 1]
{
 "attestations": [
  {
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  },
  {
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  }
 ],
 "image": {
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFile/image_signatures - 1]
{
 "attestations": [
  {
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  }
 ],
 "image": {
  "ref": "registry.io/repository/image:tag",
  "signatures": [
   {
    "chain": [
     "certificate1",
     "certificate2"
    ],
    "keyid": "keyId1",
    "sig": "signature1"
   },
   {
    "keyid": "keyId2",
    "sig": "signature2"
   }
  ],
  "source": {}
 }
}
---

[TestWriteInputFile/image_config - 1]
{
 "attestations": null,
 "image": {
  "config": {
   "Labels": {
    "io.k8s.display-name": "Test Image"
   }
  },
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFile/parent_image_config - 1]
{
 "attestations": null,
 "image": {
  "parent": {
   "config": {
    "Labels": {
     "io.k8s.display-name": "Base Image"
    }
   },
   "ref": "registry.io/repository/image/parent:tag"
  },
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFile/attestation_with_signature - 1]
{
 "attestations": [
  {
   "signatures": [
    {
     "certificate": "certificate",
     "chain": [
      "a",
      "b",
      "c"
     ],
     "keyid": "keyId",
     "metadata": {
      "k1": "v1",
      "k2": "v2"
     },
     "sig": "signature"
    }
   ],
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  }
 ],
 "image": {
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFileMultipleAttestations - 1]
{
 "attestations": [
  {
   "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
     "buildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
     "builder": {
      "id": ""
     },
     "invocation": {
      "configSource": {}
     }
    },
    "predicateType": "https://slsa.dev/provenance/v0.2",
    "subject": null
   }
  }
 ],
 "image": {
  "ref": "registry.io/repository/image:tag",
  "source": {}
 }
}
---

[TestWriteInputFile/component_with_source - 1]
{
 "attestations": null,
 "image": {
  "ref": "registry.io/repository/image:tag",
  "source": {
   "git": {
    "revision": "main",
    "url": "git.local/repository"
   }
  }
 }
}
---
